cpp/src/arrow/compute/exec/asof_join_node.cc

```diff
@@ -602,52 +828,155 @@ class AsofJoinNode : public ExecNode {
 
  public:
   AsofJoinNode(ExecPlan* plan, NodeVector inputs, std::vector<std::string> input_labels,
-               const AsofJoinNodeOptions& join_options,
+               const vec_col_index_t& indices_of_on_key,
+               const std::vector<vec_col_index_t>& indices_of_by_key, OnType tolerance,
                std::shared_ptr<Schema> output_schema);
 
+  Status InternalInit(bool must_hash, bool nullable_by_key,
+                      std::vector<std::unique_ptr<KeyHasher>> key_hashers) {
+    key_hashers_.swap(key_hashers);
```

Comment:
Why do we need to do this?

Reply:
Before this line, the `key_hashers_` field is an empty vector. The swap effectively moves the elements in the `key_hashers` argument to the `key_hashers_` field, which ensures the elements are not destructed by the caller.

Reply:
Why not `key_hashers_ = std::move(key_hashers);`?

Reply:
That works too.